internal/bzlmod/go_deps.bzl

```
@@ -276,6 +282,49 @@ _go_repository_config = repository_rule(
     },
 )
 
+def fail_on_version_conflict(version, previous, module_tag, module_name_to_go_dot_mod_label, fail_or_warn):
+    """
+    Check if duplicate modules have different versions, and fail with a useful error message if they do.
+
+    Args:
+        version: The version of the module.
+        previous: The previous module object.
+        module_tag: The module tag.
+        module_name_to_go_dot_mod_label: A dictionary mapping module paths to go.mod labels.
+        previous: The previous module object.
+    """
+
+    if not previous:
+        # no previous module, so no possible error
+        return
+
+    if not previous or version == previous.version:
+        # version is the same, skip because we won't error
+        return
+
+    # When using go.work, duplicate dependency versions are possible.
+    # This can cause issues, so we fail with a hopefully actionable error.
+    current_label = module_tag._parent_label
+
+    previous_label = previous.module_tag._parent_label
+
+    corrective_measure = "To correct this:\n 1. ensure that '{}' in all go.mod files is the same version.\n 2. in the folders where you made changes to run: go mod tidy\n 3. at the workspace root run: go work sync.".format(module_tag.path)
```

You can use `"""..."""` string literals instead, which can contain literal line breaks, to make this more readable.

```suggestion
    corrective_measure = "To correct this:\n 1. ensure that '{}' in all go.mod files is the same version.\n 2. in the folders where you made changes run: bazel run @rules_go//go -- mod tidy\n 3. at the workspace root run: bazel run @rules_go//go -- work sync.".format(module_tag.path)
```